src/psyclone/domain/lfric/psyir.py

```diff
@@ -0,0 +1,253 @@
+    TYPE = "".join(info.generic_type.title().split())
+    ARGS = ["self", "name"] + info.properties + ["interface=None"]
+    VARS = ["        self.{0} = {0}".format(var) for var in info.properties]
+    # Create the specific datatype
+    exec(
+        "class {0}DataType({1}DataType):\n"
+        "    pass\n"
+        "".format(NAME, TYPE))
+    # Create the specific symbol
+    exec(
+        "class {0}DataSymbol({1}DataSymbol):\n"
+        "    def __init__({2}):\n"
+        "{3}\n"
+        "        super({0}DataSymbol, self).__init__(\n"
+        "            name, interface=interface)\n"
+        "".format(NAME, TYPE, ", ".join(ARGS), "\n".join(VARS)))
+
+# Define LFRic field datatypes and symbols
+
+# Note, field_datatypes are no different to array_datatypes and are
+# treated in the same way. They are only separated into a different
+# list because they are used to create vector field datatypes and
+# symbols.
+
+# The first Array named-tuple argument determines the names of the
+# resultant datatype and datasymbol classes, the second argument
+# references the generic scalar type classes declared above, the third
+# argument specifies the dimensions of the array by specifying a list
+# of scalar type classes declared above, and the fourth argument
+# specifies any additional class properties that should be declared in
+# the generated datasymbol class.
+
+Array = namedtuple('Array', ["name", "scalar_type", "dims", "properties"])
+field_datatypes = [
+    Array("real field data", "lfric real scalar", ["number of unique dofs"],
+          ["fs"]),
+    Array("integer field data", "lfric integer scalar",
+          ["number of unique dofs"], ["fs"]),
+    Array("logical field data", "lfric logical scalar",
+          ["number of unique dofs"], ["fs"])]
+
+# Define all other LFRic array datatypes and symbols
+
+# TBD: #918 the dimension datatypes and their ordering is captured in
+# field_datatypes and array_datatypes but is not stored in the
+# generated classes.
+
+# TBD: #926 attributes will be constrained to certain datatypes and
+# values. For example, a function space attribute should be a string
+# containing the name of a supported function space. These are not
+# currently checked.
+
+# TBD: #927 in some cases the values of attributes can be inferred, or
+# at least must be consistent. For example, a field datatype has an
+# associated function space attribute, its dimension symbol (if there
+# is one) must be a NumberOfUniqueDofsDataSymbol which also has a
+# function space attribute and the two function spaces must be
+# the same. This is not curently checked.
+
+array_datatypes = [
+    Array("operator", "lfric real scalar",
+          ["number of dofs", "number of dofs", "number of cells"],
+          ["fs_from", "fs_to"]),
+    Array("dof map", "lfric integer scalar", ["number of dofs"], ["fs"]),
+    Array("basis function qr xyoz", "lfric real scalar",
+          ["lfric integer scalar", "number of dofs",
```

Comment:
Is the "lfric integer scalar" the thing that is either 1 or 3? Presumably we can find a name for this "thing"? i.e. what is it exactly that determines whether it's 1 or 3?

Reply:
Yes that is correct. It is 1 if it is a scalar (diff) basis function and 3 if it is a vector (diff) basis function. In `FunctionSpace.py` there are lists of function space names corresponding to scalar or vector (diff) basis functions. I guess we could have the following classes `ScalarBasisSize`, `ScalarDiffBasisSize`, `VectorBasisSize`, and `VectorDiffBasisSize` which were integers with values 1, 1, 3 and 3 respectively. What do you think? I could ask the Met Office?

Reply:
I like that suggestion. Probably best to check the precise naming with the MO. You could e.g. have "NumComponentsScalarBasis".

Reply:
I've come up with a proposed solution which subclasses the Literal node. I think we can get feedback on names separately from the Met Office and the important thing is the implementation at the moment. Note the code example still uses a scalar value for the dimension but psyir.py declares the dimensions as one would expect. I've created a separate issue #946 to replace int with Literal as a dimension argument in the PSyIR.